### ISSUE TITLE
atomWithStorage behaviour change from v1 -> v2, and documentation

### ISSUE THREAD

**Issue by wenq1** [pmndrs/jotai#1791]
## Summary

After the v2 launch, atomWithStorage by default gets its initial value from `baseAtom.onMount ()` method, making it breaking existing logic. This would be better be mentioned in the documentation which currently **at least** indicates otherwise.

## Link to reproduction

https://codesandbox.io/s/vibrant-waterfall-p96z11?file=/src/App.tsx

Click on the SETVAL button in the link and get the value updated in the local storage, then refresh the page.

With Jotai v2.0.2, after each refresh of the page, `currText` will be logged twice in `useEffect ()`, the first one `{wow: 'init'}`, and the second one `{wow: 'updated'}`.
With Jotai v1.1.13, after each refresh of the page `currText` will only be logged once, namely `{wow: 'updated'}`.

The current [documentation](https://jotai.org/docs/guides/persistence#a-simple-pattern-with-localstorage) at least indicates that `getInitialValue ()` happens immediately at the point where the atom is created, which will probably lead to confusions.

Basically, it is used to be OK to put the `currText` atom in the dependency list of `useEffect ()`. Now it has to to be placed inside that of something like `useUpdateEffect ()`. I don't know whether this change is intentional to make v2 more "async-look-and-feel" or it is just a slip.

**Comment by dai-shi** [comment]
This is an intentional breaking change to avoid hydration warning.
See also #1737 

I'll make the migration guide more explicit.

**Comment by wenq1** [comment]
thank you

**Comment by wenq1** [comment]
I missed #1737 when searching for previous issues. @dai-shi is maybe the issue template could advice a component, e.g. “[atomWithStorage] value refreshed twice in v2” so that it is more searchable, or provide a list of colored tags such as “component: atomWithStorage”, so that less duplicates are in the issues list?

**Comment by dai-shi** [comment]
> provide a list of colored tags such as “component: atomWithStorage”, so that less duplicates are in the issues list?

Would you be willing to maintain those tags??

**Comment by wenq1** [comment]
> > provide a list of colored tags such as “component: atomWithStorage”, so that less duplicates are in the issues list?
> 
> Would you be willing to maintain those tags??

Sure. Need triage access.

**Comment by dai-shi** [comment]
Added `About: atomWithStorage` label. Let me know if you need more labels.
We can chance description and color too.

**Comment by wenq1** [comment]
Ummm I don't seem to have the "Add Label" button.  Nevermind...

<img width="884" alt="image" src="https://user-images.githubusercontent.com/110781/221344077-0ae899a2-c2a6-4093-a810-2a8fda97d82e.png">

**Comment by dai-shi** [comment]
Unfortunately, you can't add them. Let me know what you need and I can add/change them.

**Comment by wenq1** [comment]
Cool. Thx
